scripts/deps/update-package-json.js

```diff
@@ -0,0 +1,154 @@
+
+/*
+ * Copyright (c) 2022 - Thoughtworks Inc. All rights reserved.
+ */
+
+const fs = require('fs');
+const path = require('path');
+const { execSync } = require('child_process');
+
+// externals
+const semver = tryRequire('semver');
+
+const WORKSPACE = path.dirname(path.dirname(__dirname));
+const SOURCES = WORKSPACE;
+
+const manifest = JSON.parse(fs.readFileSync(path.join(SOURCES, 'package.json')));
+
+const FLAGS = ['--canary', '--optional', '-h', '--help'];
+
+const opts = process.argv.slice(2).reduce((opts, arg) => {
+  if (!FLAGS.includes(arg)) {
+    usage(`Unknown argument/flag: ${arg}`);
+  }
+
+  if ('-h' === arg || '--help' === arg) {
+    usage();
+  }
+
+  if ('--canary' === arg) {
+    opts['canary'] = true;
+  }
+
+  if ('--optional' === arg) {
+    opts['update-optional-dependencies'] = true;
+  }
+
+  return opts;
+}, {});
+
+const canary = !!opts.canary;
+
+if (canary) {
+  console.warn(
+    `canary: ${canary};`,
+    'This will perform any major version upgrades as well, even for exact-version package specs',
+  );
+}
+
+function usage(msg = '') {
+  msg = (msg || '').trim();
+  const prog = path.basename(process.argv[1]);
+
+  if (msg) {
+    console.error(`[FATAL] ${msg}`);
+  }
+
+  console.error(`
+
+  USAGE: ${prog} [ OPTIONS ]
+
+  DESCRIPTION
+
+    Updates npm dependencies to the latest versions (patch or major, depending on \`--canary\`). Tree-shakes package-lock.json by
+    regenerating it from scratch.
+
+  ARGUMENTS
+
+    none
+
+  OPTIONS
+
+    -h, --help     Shows this help message.
+
+    --canary       When specified, the script will update to the latest major versions. By default (i.e., without this flag), the
+                   script only updates to the latest minor or patch version.
+
+    --optional     When specified, the script will update also update the \`optionalDependencies\` block. This may not always be
+                   desirable as we use this block to pre-download transitive optionalDependencies to satisfy builds for linux
+                   agents in the build pipeline (currently, anyway), so upgrading these might break the build if the parent package
+                   is not using the latest version.
+
+`);
+
+  process.exit(1);
+}
+
+function update(dict) {
+  interestingKeys(dict).forEach((mod) => {
+    const range = dict[mod];
+    console.log(`Checking ${mod} -- current: ${range}`);
+
+    // reverse sort to get the latest first
+    const versions = JSON.parse(execSync(`npm view ${mod} versions --json`).toString()).reverse();
+    const major = semver.major(range.replace(/^[^0-9]*/, ''));
+    const latest = semver.maxSatisfying(versions, canary ? `>=${major}` : range);
+
+    const detectModifier = range.match(/^([^0-9]+)/);
+    const modifier = detectModifier ? detectModifier[1] : '';
+    const upgraded = modifier + latest;
+
+    if (range !== upgraded) {
+      dict[mod] = upgraded;
+      console.log('  => updating', mod, 'to', upgraded);
+    }
+  });
+}
```

Comment:
This basically loops through a list of dependencies and asks the npm registry for a list of all versions.

For each package, use `semver` to determine the upgrade candidate using this logic:

- if `--canary` was specified, take the most recent version
- else, construct a `semver` constraint to find the latest version within that constraint

if the candidate is newer than that listed in `package.json`, replace the version value.